transactions/read-only/memory.md

```diff
@@ -0,0 +1,494 @@
+non-deterministic time to get the transaction result.
+
+## Proposed Solutions
+
+### Use Single WASM Interface and Compile the Same Contract Only Once in EOS-VM and EOS-VM-JIT
+
+To prevent issues of multiple WASM interfaces discussed above, it is proposed a single WASM interface is used.
+
+To avoid parsing contract code multiple times, a module is to be shared by multiple execution contexts.
+After a code is first parsed and a module is created, the module is cached, until removed from the cache when last block
+its code was used is before LIB and its linked execution context is not running. A new backend constructor
+takes a module as an input and skips parsing.
+
+To resolve the issue of potential multiple threads modifying the same global stored in the module,
+globals are moved out of module and placed in execution context.
+
+Functions `set_global` and `get_global` in `execution_context.hpp`, `emit_get_global` and `emit_set_global` in `x86_64.hpp` will
+be modified to access globals in execution context.
+
+```mermaid
+flowchart TD
+   subgraph module[module]
+   end
+   subgraph wasmif
+      subgraph backend_1[backend]
+         subgraph execution_ctx_1[execution context]
+            direction TB
+            subgraph LM1[linear memory]
+            end
+            subgraph Stack1[stack]
+            end
+            LM1 ~~~ Stack1
+         end
+      end
+
+      subgraph backend_2[backend]
+         subgraph execution_ctx_2[execution context]
+            direction TB
+            subgraph LM2[linear memory]
+            end
+            subgraph Stack2[stack]
+            end
+            LM2 ~~~ Stack2
+         end
+      end
+   end   
+   backend_1 ~~~ backend_2
+   module --> backend_1
+   module --> backend_2
+   
+```
+
+### Reduce Memory Slice Counts
+To conserve virtual memory to support more threads, We need to reduce the threshold where EOS VM OC transitions
+from mirroring to `mprotect()`.
+
+We plan to
+- Gather memory usage from existing contracts.
+  * Download a January 2022 snapshot from the mainnet.
+  * Modify `libraries/chain/webassembly/runtimes/eos-vm-oc/gs_seg_helpers.c/eos_vm_oc_grow_memory()` to keep track the largest number of linear memory the current code uses, and modify `eosvmoc::executor::execute()` to report the larget number at the end of the action execution.
+  * Start from the snapshot and sync with a mainnet peer node.
+  * Find max and 95 percentile of number of linear memory pages used by all codes. This is to make sure vast majority of executions do not need using `mprotect`.
+- Add a private compile option defining the threshold number of pages where the transition between the two approaches occurs.
+  * In `chain/CMakefile`, set a private variable `max_num_slices` with the value of the 95 percentile, and pass it to C++ code using `add_definitions`.
+  * Use `max_num_slices` to set up memory slices in `memory::memory()`.
```

Comment:
I'd like to see https://github.com/AntelopeIO/leap/issues/1256#issuecomment-1587604359 considered instead. Possibly even left at 33MB for the main thread due to the EVM numbers you've discovered.